### PR TITLE
drivers: mpsl:  Add flash sync driver

### PR DESCRIPTION
**Pull request by eugmes**
This PR adds flash sync support for the `soc_flash_nrf` driver using MPSL timeslot API. It also removes the flash driver provided by `bt_ll_softdevice` and replaces it by `soc_flash_nrf` with MPSL sync enabled.

### CODE REVIEW

**Comment by eugmes** [comment]
An easier to read version is here: https://github.com/tejlmand/fw-nrfconnect-nrf/pull/2

**Comment by eugmes** [comment]
> Looks like BLE connections / security are broken. Can't establish connection with nrf_desktop applications.
> Security fails:
> ble_latency: Security establishment failed

Are you able to establish connections when changes introduced in this PR are removed, with just the changes from the upmerge branch?

**Comment by wiminordic** [comment]
> > Looks like BLE connections / security are broken. Can't establish connection with nrf_desktop applications.
> > Security fails:
> > ble_latency: Security establishment failed
> 
> Are you able to establish connections when changes introduced in this PR are removed, with just the changes from the upmerge branch?

Yes.

**Comment by eugmes** [comment]
@wiminordic It should not trigger MPSL asserts anymore at least. Connection is still unreliable though.

**Comment by eugmes** [comment]
It seems there are some problems with spinlocks while running the nrf_desktop app:

image

**Comment by eugmes** [comment]
> It seems there are some problems with spinlocks while running the nrf_desktop app:

The address points to `sem.c:lock`

**Comment by eugmes** [comment]
> It seems there are some problems with spinlocks while running the nrf_desktop app:

Not my problem: https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/latest%2Fsub%2Fdesktop52_verification/detail/master/4902/pipeline

**Comment by eugmes** [comment]
@wiminordic Flash operations should not timeout anymore with the nrf_desktop app. There are problems in the upmerge branch however that may cause asserts.

**Comment by wiminordic** [comment]
> @wiminordic Flash operations should not timeout anymore with the nrf_desktop app. There are problems in the upmerge branch however that may cause asserts.

Let's see how the tests go. We can wait for the upmerge to go green, then rebasing this PR to make sure it is not causing problems.

**Comment by wiminordic** [comment]
ASSERTION FAIL [z_spin_lock_valid(l)] @ WEST_TOPDIR/zephyr/include/spinlock.h:92	Recursive spinlock 0x20007960
This issue was not present for upmerge last time it was tested.

**Comment by eugmes** [comment]
> ASSERTION FAIL [z_spin_lock_valid(l)] @ WEST_TOPDIR/zephyr/include/spinlock.h:92 Recursive spinlock 0x20007960
> This issue was not present for upmerge last time it was tested.

I've seen that issue (https://github.com/nrfconnect/sdk-nrf/pull/3003#issuecomment-702234132), but I was pretty sure that I've also seen it in the upmerge branch (https://github.com/nrfconnect/sdk-nrf/pull/3003#issuecomment-702364563). I now see that the link I posted does not show this problem, but some other problem with MPSL. It looks like something enables interrups when it shouldn't...

**Comment by ru-fu** [comment]
No need for a doc review for the current changes.
But should there be documentation for the new driver?

**Comment by eugmes** [comment]
@wiminordic Recursive spinlock assert is gone now.

**Comment by wiminordic** [comment]
No crash anymore, but Dongle throws a ble_scan error log: 
<err> ble_scan: Problem storing subscribed_peers (err -22)
@pdunaj could that be something Desktop related or something lower level?

**Comment by pdunaj** [comment]
> ble_scan: Problem storing subscribed_peers (err -22)

The error is return from `settings_save_one` -22 is `EINVAL`.
`settings_fcb_save` can return `EINVAL` when `(val_len > 0 && value == NULL)` but that would be strange - it works on master. Maybe it is `EINVAL` from settings_fcb_save_priv` (why would one use `EINVAL` here why not returning the error itself @nvlsianpu )? `fcb_append` calls `flash_area_write` at some point but since any error in here will be `EINVAL` it's hard to tell what happened.

I suspect this is a problem of this change as we don't see problems on master.

**Comment by eugmes** [comment]
@wiminordic  @pdunaj This flash sync driver would return `-EINVAL` if call to `mpsl_timeslot_request` has failed (https://github.com/nrfconnect/sdk-nrf/pull/3003/files#diff-5d0cb4a9ddfd16f8fbd406f5e2f18454R182), but it also would also log an error message. I don't see it in the log.

**Comment by wiminordic** [comment]
The ble_scan error is the only error logged on RTT by the Dongle. It occurs after the second device connects to it. First connection throws no errors.

File: drivers/mpsl/flash_sync/flash_sync_mpsl.c [hunk 1]
```diff
@@ -179,7 +179,7 @@ int nrf_flash_sync_exe(struct flash_op_desc *op_desc)
 	if (ret < 0) {
 		LOG_ERR("mpsl_timeslot_request failed: %d", ret);
 		mpsl_timeslot_session_close(_context.session_id);
-		return -EINVAL;
+		return -EHOSTDOWN;
```
Comment (eugmes) [inline drivers/mpsl/flash_sync/flash_sync_mpsl.c:182]: This is for testing only.

- [ ] Undo that

**Comment by eugmes** [comment]
The reported error code did not change after modifying the MPSL flash sync driver: https://jenkins-ncs.nordicsemi.no/blue/organizations/jenkins/latest%2Fsub%2Fdesktop52_verification/detail/master/5023/pipeline, so it is probably reported by ether the setting subsystem or by Zephyr's NRF flash driver.

**Comment by eugmes** [comment]
There were no additional errors logged even after adding error logging to the nrf flash driver. The errors are either filtered out or they come from the setting subsystem.

**Comment by eugmes** [comment]
`desktop52_verification` is passing now. The problem was with the timeslot priority escalation.

File: drivers/bt_ll_softdevice/CMakeLists.txt [hunk 2]
```diff
@@ -4,5 +4,4 @@
 # SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
```
Comment (eugmes) [inline drivers/bt_ll_softdevice/CMakeLists.txt:4]: The SDC flash driver is removed so the new one could be tested. I think it is possible to have both drivers present at the same time.

**Comment by eugmes** [comment]
https://github.com/nrfconnect/sdk-nrf/pull/3067 should be reviewer instead.

**Pull request closed without merging**